python/dgl/runtime/spmv.py

```diff
@@ -165,236 +130,86 @@ def build_adj_matrix_graph(graph):
+    nbits = gidx.bits_needed()
+    forward_map = utils.to_nbits_int(eid[forward.tousertensor()], nbits)
+    backward_map = utils.to_nbits_int(eid[backward.tousertensor()], nbits)
+    forward_map = F.zerocopy_to_dgl_ndarray(forward_map)
+    backward_map = F.zerocopy_to_dgl_ndarray(backward_map)
+    edge_map = utils.CtxCachedObject(
+        lambda ctx: (nd.array(forward_map, ctx=ctx),
+                     nd.array(backward_map, ctx=ctx)))
+    return utils.CtxCachedObject(gidx.get_immutable_gidx), edge_map, nbits
 
-def build_inc_matrix_dst(dst, reduce_nodes):
-    """Build incidence matrix using only edge destinations.
 
-    The incidence matrix is of shape (n, m), where n=len(reduce_nodes), m=len(dst).
-    The nnz is equal to len(dst).
-
-    Examples
-    --------
-    Five edges. Two edges point to node 1; one points to node 3;
-    two point to node 4. There are five nodes in the final
-    target dimension (0~4), where node 0 and 2 are two 0-deg nodes.
-    >>> dst = [1, 1, 3, 4, 4]
-    >>> reduce_nodes = [0, 1, 2, 3, 4]
-    >>> build_inc_matrix_dst(dst, reduced_nodes)
-    tensor([[0, 0, 0, 0, 0],
-            [1, 1, 0, 0, 0],
-            [0, 0, 0, 0, 0],
-            [0, 0, 1, 0, 0],
-            [0, 0, 0, 1, 1]], shape=(5, 5))
+def build_gidx_and_mapping_block(graph, block_id, edge_tuples=None):
+    """Build immutable graph index and mapping for node flow
 
     Parameters
     ----------
-    dst : utils.Index
-        The edge destinations.
-    reduce_nodes : utils.Index
-        The nodes to reduce messages, which will be target dimension
-        of the incmat. The nodes include unique(dst) and zero-degree-nodes.
+    graph : NodeFlow
+        The NodeFlow
+    block_id : int
+        the block Id
+    edge_tuple :  tuple of three utils.Index
+        A tuple of (u, v, eid)
 
     Returns
     -------
-    utils.CtxCachedObject
-        Get be used to get incidence matrix on the provided ctx.
-    utils.Index
-        A index for data shuffling due to sparse format change. Return None
-        if shuffle is not required.
+    graph : utils.CtxCachedObject
+        Function that generates a immutable graph index on given context
+    edge_map : utils.CtxCachedObject
+        Function that generates forward and backward edge mapping on given
+        context
+    nbits : int
+        Number of ints needed to represent the graph
     """
-    eid = utils.toindex(F.arange(0, len(dst)))
-    return build_inc_matrix_eid(len(eid), eid, dst, reduce_nodes)
+    if edge_tuples is None:
+        u, v, eid = graph.block_edges(block_id, remap=True)
+        u = utils.toindex(u)
+        v = utils.toindex(v)
+        eid = utils.toindex(eid)
+    else:
+        u, v, eid = edge_tuples
+    num_nodes = max(graph.layer_size(block_id), graph.layer_size(block_id + 1))
+    gidx, edge_map, nbits = build_gidx_and_mapping_uv((u, v, eid), num_nodes)
+    return gidx, edge_map, nbits
```

Comment:
please slice block csr directly.

Reply:
Will do in next PR. There are some subtlety I need to carefully think about.